aws/src/main/java/org/apache/iceberg/aws/glue/DynamoLockManager.java

```diff
@@ -0,0 +1,308 @@
+          .build());
+      return true;
+    } catch (ResourceNotFoundException e) {
+      return false;
+    }
+  }
+
+  private void checkTableActive(String tableName) {
+    try {
+      DescribeTableResponse response = dynamo.describeTable(DescribeTableRequest.builder()
+          .tableName(tableName)
+          .build());
+      TableStatus currentStatus = response.table().tableStatus();
+      if (!currentStatus.equals(TableStatus.ACTIVE)) {
+        throw new IllegalStateException(String.format("Dynamo table %s is not active, current status: %s",
+            tableName, currentStatus));
+      }
+    } catch (ResourceNotFoundException e) {
+      throw new IllegalStateException(String.format("Cannot find Dynamo table %s", tableName));
+    }
+  }
+
+  @Override
+  public void initialize(Map<String, String> properties) {
+    super.initialize(properties);
+    this.dynamo = AwsClientFactories.from(properties).dynamo();
+    this.lockTableName = properties.get(CatalogProperties.LOCK_TABLE);
+    Preconditions.checkNotNull(lockTableName, "DynamoDB lock table name must not be null");
+    ensureLockTableExistsOrCreate();
+  }
+
+  @Override
+  public void acquire(LockId lockId) {
+    try {
+      Tasks.foreach(lockId)
+          .throwFailureWhenFinished()
+          .retry(Integer.MAX_VALUE - 1)
+          .exponentialBackoff(acquireIntervalMs(), acquireIntervalMs(), acquireTimeoutMs(), 1)
+          .onlyRetryOn(
+              ConditionalCheckFailedException.class,
+              ProvisionedThroughputExceededException.class,
+              TransactionConflictException.class,
+              RequestLimitExceededException.class,
+              InternalServerErrorException.class)
+          .run(this::acquireOnce);
+    } catch (RuntimeException e) {
+      throw new LockNotAcquiredException("Fail to acquire lock " + lockId, e);
+    }
+  }
+
+  @VisibleForTesting
+  void acquireOnce(LockId lockId) {
+    GetItemResponse response = dynamo.getItem(GetItemRequest.builder()
+        .tableName(lockTableName)
+        .key(toKey(lockId))
+        .build());
+
+    if (!response.hasItem()) {
+      dynamo.putItem(PutItemRequest.builder()
+          .tableName(lockTableName)
+          .item(toNewItem(lockId))
+          .conditionExpression(CONDITION_LOCK_ENTITY_NOT_EXIST)
+          .build());
+    } else {
+      Map<String, AttributeValue> currentItem = response.item();
+
+      try {
+        Thread.sleep(Long.parseLong(currentItem.get(COL_LEASE_DURATION_MS).n()));
+      } catch (InterruptedException e) {
+        throw new LockNotAcquiredException(
+                String.format("Fail to acquire lock %s, interrupted during sleep", lockId), e);
+      }
+
+      Map<String, AttributeValue> expressionValues = Maps.newHashMap();
+      expressionValues.put(":eid", AttributeValue.builder().s(lockId.entityId()).build());
+      expressionValues.put(":vid", AttributeValue.builder().s(currentItem.get(COL_VERSION).s()).build());
```

Comment:
Can you also store the interval in the row? That way, the process that holds the lock shares its expectation with other processes. If you don't do that, then another process could use a wait interval shorter than the lock holder's heartbeat interval and incorrectly unlock. So the expected wait interval should come from the process that currently holds the lock, not the process attempting to get it.

Reply:
I don't think the situation you describe would happen. The heartbeat timeout duration is already stored (L270), and the lock acquisition process would wait for the stored heartbeat timeout duration before trying to acquire (L211). 

This means the total acquisition time might exceed `lock.acquire-timeout-ms`, because it will first wait for the heartbeat timeout duration and then check if acquisition time has timed out or not. And the `lock.acquire-interval-ms` can be very very short, and this lock mechanism would still work.

Maybe you are looking at the wrong line. The `expressionValues` here are only used to check against `CONDITION_LOCK_ENTITY_NOT_EXIST_OR_VERSION_MATCH`, the actual data stored is at `toNewItem` at L270.